### PR TITLE
(py-)onnx: new version 1.13.0, 1.13.1 (for py-torch@2) 

### PR DESCRIPTION
**Pull request by wdconinc**
ONNX has a new version 1.13.1 which is required for py-torch +onnx_ml.
This version adds python 3.11 support, Mac M1/M2 support. No build
system changes necessary.

- https://github.com/onnx/onnx/releases/tag/v1.13.0 (main changes)
- https://github.com/onnx/onnx/releases/tag/v1.13.1 (bugfixes)
- https://github.com/onnx/onnx/commits/main/CMakeLists.txt (top 6
  commits are on top of the last 1.12 release)
- https://github.com/onnx/onnx/blob/v1.13.1/requirements.txt
- https://github.com/onnx/onnx/blob/v1.13.1/requirements-release.txt

Because this new version of ONNX depends on `protobuf@3.20.2:`, we need to add two new patch versions of `protobuf` and `py-protobuf`, and we change the preferred version to the most recent patch version 3.20.3. No build system changes necessary.
- https://github.com/protocolbuffers/protobuf/compare/v3.20.1...v3.20.3

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-onnx/package.py [hunk 1]
```diff
@@ -42,8 +44,12 @@ class PyOnnx(PythonPackage):
     depends_on("py-protobuf@3.12.2:", type=("build", "run"), when="@1.11.0")
     depends_on("protobuf@3.12.2:3.20.1", when="@1.12.0:")
     depends_on("py-protobuf@3.12.2:3.20.1", type=("build", "run"), when="@1.12.0:")
+    depends_on("protobuf@3.20.2:", when="@1.13.0:")
+    depends_on("py-protobuf@3.20.2:", type=("build", "run"), when="@1.13.0:")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-onnx/package.py]:
```suggestion
    depends_on("protobuf@3.12.2:3.20.1", when="@1.12.0")
    depends_on("py-protobuf@3.12.2:3.20.1", type=("build", "run"), when="@1.12.0")
    depends_on("protobuf@3.20.2:3", when="@1.13.0:")
    depends_on("py-protobuf@3.20.2:3", type=("build", "run"), when="@1.13.0:")
```

File: var/spack/repos/builtin/packages/py-onnx/package.py [hunk 2]
```diff
@@ -42,8 +44,12 @@ class PyOnnx(PythonPackage):
...
     depends_on("py-numpy", type=("build", "run"))
     depends_on("py-numpy@1.16.6:", type=("build", "run"), when="@1.8.1:")
+    depends_on("py-numpy@1.21.5:", type=("build", "run"), when="@1.13:")
+    depends_on("py-numpy@1.23.2:", type=("build", "run"), when="@1.13: ^python@3.11:")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-onnx/package.py]: According to `requirements.txt`, it's still numpy >= 1.16.6

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-onnx/package.py (hunk 2 above)
Comment (wdconinc) [reply var/spack/repos/builtin/packages/py-onnx/package.py]: Right, was looking at main, and https://github.com/onnx/onnx/pull/5095

File: var/spack/repos/builtin/packages/py-onnx/package.py (hunk 1 above)
Comment (wdconinc) [reply var/spack/repos/builtin/packages/py-onnx/package.py]: Good catch, thanks!

File: var/spack/repos/builtin/packages/py-onnx/package.py (hunk 2 above)
Comment (wdconinc) [reply var/spack/repos/builtin/packages/py-onnx/package.py]:
```suggestion
    depends_on("py-numpy@1.16.6:", type=("build", "run"), when="@1.8.1:")
```

**Comment by wdconinc** [comment]
Draft since hung up on protobuf 3.22 c++14 requirement due to abseil-cpp, and default onnx c++11, i.e. fails compilation without patch of cmakelists for onnx.

**Comment by adamjstewart** [comment]
I had to hack something like this for paraview in #36711, you might be able to use the same approach.

**Review by adamjstewart — DISMISSED** [review]

**Comment by wdconinc** [comment]
Yeah, https://github.com/onnx/onnx/issues/4992 is what I was looking at.

**Comment by wdconinc** [comment]
I'll put the c++14 stuff in a separate PR.

**Comment by wdconinc** [comment]
Marked as ready. This builds fine with:
```console
$ spack find -lxv
==> In environment onnx
==> Root specs
------- onnx@1.13.1  ------- py-onnx@1.13.1

==> Installed packages
-- linux-ubuntu23.04-skylake / gcc@12.2.0 -----------------------
m3g6jqh onnx@1.13.1~ipo build_system=cmake build_type=RelWithDebInfo generator=ninja     w6o6yxw py-onnx@1.13.1 build_system=python_pip
6eicdvn protobuf@3.20.3~ipo+shared build_system=cmake build_type=Release generator=make  o4dtnrh py-protobuf@3.20.3+cpp build_system=python_pip
==> 4 installed packages
```

What is not working yet is building with `protobuf@3.22` because there is no equivalent `py-protobuf@3.22`. But that only happens when a user willfully adds `protobuf@3.22` to an environment or spec.

Oh, and
@spackbot fix style

File: var/spack/repos/builtin/packages/protobuf/package.py [hunk 3]
```diff
@@ -81,6 +83,7 @@ class Protobuf(CMakePackage):
     # https://github.com/protocolbuffers/protobuf/issues/11828#issuecomment-1433557509
     depends_on("abseil-cpp@20230125:", when="@3.22:")
     depends_on("zlib")
+    depends_on("abseil-cpp", when="@3.22:")
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/protobuf/package.py]: This was pulled in through a rebase.
```suggestion
```

File: var/spack/repos/builtin/packages/py-protobuf/package.py [hunk 4]
```diff
@@ -24,6 +24,12 @@ class PyProtobuf(PythonPackage):
     version("4.21.9", sha256="61f21493d96d2a77f9ca84fefa105872550ab5ef71d21c458eb80edcf4885a99")
     version("4.21.7", sha256="71d9dba03ed3432c878a801e2ea51e034b0ea01cf3a4344fb60166cb5f6c8757")
     version("4.21.5", sha256="eb1106e87e095628e96884a877a51cdb90087106ee693925ec0a300468a9be3a")
+    version(
+        "3.20.3",
+        sha256="2e3427429c9cffebf259491be0af70189607f365c2f41c7c3764af6f337105f2",
+        preferred=True,
+    )
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/py-protobuf/package.py]: No more preferred version.
```suggestion
    version("3.20.3", sha256="2e3427429c9cffebf259491be0af70189607f365c2f41c7c3764af6f337105f2")
```

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**